testplan/runners/pools/base.py

```diff
@@ -441,16 +535,25 @@ def _handle_taskresults(self, worker, request, response):
         for task_result in request.data:
             uid = task_result.task.uid()
             worker.assigned.remove(uid)
-            if worker not in self._workers_last_result:
-                self._workers_last_result[worker] = time.time()
+            self._workers_last_result.setdefault(worker, time.time())
             self.logger.test_info(
                 "De-assign {} from {}".format(task_result.task, worker)
             )
 
             if self.should_reschedule(self, task_result):
-                if self.task_assign_cnt[uid] >= self.cfg.task_retries_limit:
+                self.logger.test_info(
+                    "Will reschedule %(task)s due to `should_reschedule` "
+                    "cfg option of %(pool)s",
+                    {"task": task_result.task, "pool": self,},
+                )
+                self.unassigned.append(uid)
+                continue
+
+            if self.should_rerun(self, task_result):
+                if self.task_assign_cnt[uid] > self.cfg.task_retries_limit:
```

Comment:
should_rerun already checked against task's rerun limit, the task_retries_limited is checked in taskpull handler.

In a treadmill pool, if a task is scheduled out to a worker, and in the worker's process pool, it has already retried 3 time (but worker died every time with no result returned), and finally the task is returned back to treadmill, treadmill need to reset its retry number and rerun this task again. At least this is how its handled in ets/python.